# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/client/CanvasClient.h
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 gfx/layers/client/CanvasClient.h
--- a/gfx/layers/client/CanvasClient.h	Wed Jan 08 21:52:03 2020 +0000
+++ b/gfx/layers/client/CanvasClient.h	Wed Jan 08 22:19:14 2020 +0000
@@ -23,6 +23,8 @@
 #include "mozilla/gfx/Point.h"  // for IntSize
 #include "mozilla/gfx/Types.h"  // for SurfaceFormat
 
+class nsICanvasRenderingContextInternal;
+
 namespace mozilla {
 namespace layers {
 
@@ -31,6 +33,7 @@
 class CompositableForwarder;
 class ShadowableLayer;
 class SharedSurfaceTextureClient;
+class OOPCanvasRenderer;
 
 /**
  * Compositable client for 2d and webgl canvas.
@@ -48,7 +51,8 @@
     CanvasClientSurface,
     CanvasClientGLContext,
     CanvasClientTypeShSurf,
-    CanvasClientAsync,  // webgl on workers
+    CanvasClientAsync,    // webgl on workers
+    CanvasClientTypeOOP,  // webgl in remote process
   };
   static already_AddRefed<CanvasClient> CreateCanvasClient(
       CanvasClientType aType, CompositableForwarder* aFwd, TextureFlags aFlags);
@@ -167,7 +171,9 @@
  public:
   CanvasClientBridge(CompositableForwarder* aLayerForwarder,
                      TextureFlags aFlags)
-      : CanvasClient(aLayerForwarder, aFlags), mLayer(nullptr) {}
+      : CanvasClient(aLayerForwarder, aFlags), mLayer(nullptr) {
+    mIsAsync = true;
+  }
 
   TextureInfo GetTextureInfo() const override {
     return TextureInfo(CompositableType::IMAGE);
@@ -185,6 +191,34 @@
   ShadowableLayer* mLayer;
 };
 
+/**
+ * Used for WebGL instances that perform all composition in the host process.
+ */
+class CanvasClientOOP final : public CanvasClient {
+ public:
+  CanvasClientOOP(CompositableForwarder* aLayerForwarder, TextureFlags aFlags);
+  ~CanvasClientOOP();
+
+  TextureInfo GetTextureInfo() const override {
+    return TextureInfo(CompositableType::IMAGE);
+  }
+
+  virtual void Update(gfx::IntSize aSize,
+                      ShareableCanvasRenderer* aCanvasRenderer,
+                      wr::RenderRoot aRenderRoot) override;
+
+  virtual void UpdateAsync(AsyncCanvasRenderer* aRenderer) override {
+    MOZ_ASSERT_UNREACHABLE("Illegal to call UpdateAsync on CanvasClientOOP");
+  }
+
+  void SetLayer(ShadowableLayer* aLayer, OOPCanvasRenderer* aRenderer);
+
+ protected:
+  nsICanvasRenderingContextInternal* mCanvasContext = nullptr;
+  ShadowableLayer* mLayer = nullptr;
+  CompositableHandle mHandle;
+};
+
 }  // namespace layers
 }  // namespace mozilla
 